### ISSUE TITLE
128-bit integers

### ISSUE THREAD

**Issue by seibert** [numba/numba#2600]
Can we add support for int128/uint128?  I think LLVM can generate code for this type on x86_64 (probably not ARM), but we would need to check.  Also need to ensure that nothing ever upcasts to 128-bit unless one of the operands is 128-bit.  (i.e. int64 * int64 does **not** produce int128)

**Comment by jason-s** [comment]
Thanks for considering! The numpy package maintainers don't want to touch this right now (https://github.com/numpy/numpy/issues/9992) which makes it somewhat weird as np.array may not be able to handle it the same way that smaller integer types are handled.

But for scalar values, it would definitely be helpful in some applications.

**Comment by seibert** [comment]
Interesting that NumPy got float16 before int128...

**Comment by jason-s** [comment]
A tangential topic: is it any easier to support fixed-precision numerical types, with bit widths fixed at object construction time, than arbitrary-precision numerical types?

In other words, if I wanted to use 48-bit integers or 192-bit integers or 384-bit integers, but the math is "C-like" and always yields the same output type (192x192 multiply yields 192-bit output).

**Comment by seibert** [comment]
Not really.  LLVM IR supports describing arbitrary width integers (internally LLVM can generate 65-bit integer operations, for example), but the backend that produces machine code has to figure out how to emit an implementation.  For a few specific type sizes, LLVM emits calls to functions in a "compiler_rt" library to do the higher precision work, but I don't think there is code generation for generic (but fixed) width integers.

**Comment by svenrahmann** [comment]
Any news on this? For hashing, it would be very useful to be able to compute mulmod(x,y,k) = (x*y) mod k, where x and y are uint64, the intermediate product is uint128, and the final result after mod is uint64 again. One has to ensure that the intermediate result is not implicitly taken mod 2**64 by constraining the intermediate type to uint64, so uint128 is required for this one value.
